### PR TITLE
Propagate type condition from nested fragments to parent

### PR DESCRIPTION
**Pull request by kyri-petrou**
Fix for sub-issue for #2076.

In short, when we have nested fragments with type conditions, we propagate the subtype selection to the top-level fragment.

@guymers, would you by any chance be able to publish this branch locally and test it against your schema to make sure I didn't accidentally cause something else to start failing. It shouldn't, but there might be some edge-case that we didn't account for

### CODE REVIEW

**Comment by kyri-petrou** [comment]
@ghostdogpr PS: This change means that we no longer need to check for duplicates within the Executor (changes in the previous PR), but I'd rather play it safe for this release (even if it's ever so slightly less performant) and perhaps change it in the future if we're confident we can remove it

**Comment by guymers** [comment]
This PR passes the test that I added for the issue on 2.5.0 and was failing on 2.5.1 due to the extra fields.

All our existing tests pass (but they pass on all versions).

File: core/src/main/scala/caliban/execution/Field.scala [hunk 1]
```diff
@@ -186,47 +185,60 @@ object Field {
...
+              condition = subtypeNames0
+            ).foreach { case (f, c) =>
+              if (c.exists(isSubsetCondition)) addField(f, c)
+              else addField(f, typeName)
+            }
           }
       }
-      map.values().asScala.toList
+      val builder = List.newBuilder[(Field, Option[String])]
+      map.forEach { case ((_, cond), field) => builder += ((field, cond)) }
```
Comment (ghostdogpr) [inline core/src/main/scala/caliban/execution/Field.scala:236]: Is this zio-prelude's forEach? There's no `foreach`?

**Review by ghostdogpr — APPROVED** [review]

File: core/src/main/scala/caliban/execution/Field.scala (hunk 1 above)
Comment (kyri-petrou) [reply core/src/main/scala/caliban/execution/Field.scala:236]: No this is Java's `LinkedHashMap#forEach`

Comment (ghostdogpr) [reply core/src/main/scala/caliban/execution/Field.scala:236]: Aha, was confused by the casing
